### ISSUE TITLE
Training becomes unstable after ~9k opt_steps (loss jump + grad_norm explosion). Could it be small per-GPU BN batch or too large LR?

### ISSUE THREAD

**Issue by rongrongbian67-dotcom** [shihao1895/SpatialActor#5]
thanks for releasing SpatialActor！
I’m seeing a clear instability during training: Before ~8k opt_steps: loss curves look normal. Around opt_step ~9k–10k: total loss suddenly jumps (roughly ~4–6 -> ~11–13), grad_norm sharply increases and keeps growing (reaches ~1e2–1e3) , rot_z, grip_loss, and collision_loss keep increasing after the jump, while rot_x/rot_y continue decreasing.

I attach the plots for: total loss, rotation losses, grip/collision loss, and grad_norm, and I stopped this run at ~20k opt_steps (did not reach 40k). 

<img width="950" height="680" alt="Image" src="https://github.com/user-attachments/assets/02d5fba5-ca86-48e9-ac73-6bb15fac7531" />

<img width="954" height="678" alt="Image" src="https://github.com/user-attachments/assets/cf74eef1-05fb-4a13-97f3-81a0977e2872" />

<img width="930" height="684" alt="Image" src="https://github.com/user-attachments/assets/e9acbd6e-cf3f-4bd8-abf4-710a0116ef43" />

<img width="1175" height="930" alt="Image" src="https://github.com/user-attachments/assets/372a176d-4885-4ce5-ac66-4113758ff5de" />

Based on this pattern, here are two things I suspect might be contributing:
1. BatchNorm instability due to very small per-GPU micro-batch
My per-GPU batch is only 3. Effective batch is 192 via grad accumulation.
2. Peak learning rate might be too large for my setup
Peak lr is 0.0024 (after 2000-step warmup), then cosine decay. Could lr=2.4e-3 be too aggressive under my hardware/batch regime (2 GPUs, per-GPU batch 3, accumulation 32)?

Also, if there are other common pitfalls I’m missing , I’d really appreciate any pointers.

Below is my full training configuration . I’m looking forward to your views and any recommended fixes for stable training.

Training config
- GPUs: 2× RTX 3090 (24GB each)
- per-GPU batch size: 3
- grad accumulation: 32
- effective batch size: 3 * 2 * 32 = 192
- total opt_steps: target 40k (stopped at ~20k)
- warmup: 2000 opt_steps
- peak lr: 0.0024, then cosine decay
- gradient clipping: enabled (early attempts had NaN around ~1500 steps without clipping)
- dataset/tasks: RLBench 18 tasks (train split)

**Comment by shihao1895** [comment]
Hi, thanks for trying SpatialActor.

Batch size=3 per GPU is very likely a key factor. Our recommended setup is 8×4090/A100 with per-GPU batch size ≥8. I’m not sure whether gradient accumulation helps here.

For LR, the default lr=1.25e-5 in `spatial_actor/configs/spact.yaml` is defined as a per-GPU, per-batch LR. So if you use fewer GPUs or a smaller per-GPU batch, the total LR decreases automatically.

We’d suggest keeping this lr=1.25e-5 in `spatial_actor/configs/spact.yaml` unchanged and trying to increase the GPU batch size to 8. With only 2 GPUs, we’re not sure whether the results can be fully reproduced, but it’s still worth a try.

**Comment by rongrongbian67-dotcom** [comment]
Thanks for your reply. I also wanted to ask: after I change the batch size to 8, do I still need to use gradient accumulation? If I don’t, my effective total batch size won’t reach 8 × 24 = 192. For example, with 2 GPUs, the total batch size would only be 2 × 8 = 16. Would this affect the training results?

Also: in your training setup with 8× RTX 4090 / A100 GPUs, how much GPU memory did each have? My RTX 3090 doesn’t seem to be able to increase the batch size to 8 :（.

**Comment by shihao1895** [comment]
Hi,

I’m not sure whether reducing the number of GPUs or enabling/disabling gradient accumulation will affect the training results, so you may try. 

RTX 4090 has 24GB memory, which should be similar to RTX 3090. On our side, 24GB GPUs can run with batch size = 8. If that still does not fit on your 3090, you can slightly reduce the batch size to make it stable.

I’ve also started a run on 2× 4090 GPUs and will share the results in a few days.

**Comment by rongrongbian67-dotcom** [comment]
Hello, may I ask how the loss is looking on your side with 2× RTX 4090 when doing full training? On my side, with 2× RTX 3090, the estimated training time is over 200 hours, which is too long for me. I might choose to rent GPUs, so I’d like to understand what configuration would be a good choice.

Specifically, I want to know whether training with 2× RTX 4090 can work well. If it‘s working, I have a few more questions: do you use gradient accumulation, and what would the total estimated training time be in hours? Or can you share the training configuration with me please?

Thanks a lot for your time and help!

**Comment by shihao1895** [comment]
Hi,

I reproduced the training with 2× RTX 4090, and the results on my side are as follows:
- The total training time with 2 GPUs is about 110 hours
- Using only 2 GPUs, the result on RLBench is 85.11, which is about 2 points lower than the full setting
- I did not use gradient accumulation

**Loss curve**

<img width="1660" height="991" alt="Image" src="https://github.com/user-attachments/assets/943c0101-be5b-41f1-9b1d-9bdfd3049728" />

**LR curve**

<img width="1652" height="982" alt="Image" src="https://github.com/user-attachments/assets/a4a61ed2-379b-4b65-9c79-0b96486cbdcb" />

**Training configuration**
```
exp_id: base
tasks: all
bs: 8
epochs: 50
num_workers: 8
sample_distribution_mode: task_uniform
resume: last
lr: 1.25e-5
warmup_steps: 2000
optimizer_type: lamb
lr_cos_dec: True
transform_augmentation_xyz: [0.125, 0.125, 0.125]
transform_augmentation_rpy: [0.0, 0.0, 45.0]
amp: True
bnb: True
lambda_weight_l2: 1e-4
place_with_mean: False
img_aug: 0.0
noise_type: none
model:
  norm_corr: True
  rend_three_views: True
  img_size: 224
  xops: True
  st_wpt_loc_inp_no_noise: True
  im_channels: 128
  proprio_cat_dim: 64
  lang_dim: 512
  reenc_text: False
  sem_enc_type: CLIP-RN101
  geo_enc_type: RN50
  dep_exp_type: 'DA-vitb'
  dep_exp_path: '/data/ckpt/spact/pretrained'
  align_loss: 0.0
```

**Comment by rongrongbian67-dotcom** [comment]
Thank you very much for sharing. May I ask how much memory your A100 GPUs had for training (40GB or 80GB), and what your detailed training configuration was? I’d like to try using 2× A100 GPUs as well.

**Comment by shihao1895** [comment]
Hi,

Our A100 GPUs were the 80GB version. With 2× A100 80GB, you can try increasing the batch size to 24 and keep the other settings unchanged first.

**Comment by rongrongbian67-dotcom** [comment]
Hello, I'd like to ask: when you trained for 50 epochs on 2× RTX 4090, what was the total number of training steps?

From the paper, I understand that the original setup used 8× A100 with a batch size of 8×24, 50 epochs, and about 40k steps. I'm now thinking about using 4× A100 with batch size 4×24. If I keep the default config of 50 epochs and train_iteration = 16*10000, then the total number of training steps would become:

50 * 160000 / (4 * 24) = 83300 steps

Does that sound reasonable?

Or would it make more sense to modify train_iter so that, while still keeping 50 epochs, the total training steps stay around 40k instead? Which approach do you think is more practical?

Thanks in advance for your support!

**Comment by shihao1895** [comment]
Hi, 

For 2× RTX 4090 with per-GPU bs=8, it’s about 500k iterations for 50 epochs.
For 4× A100 with per-GPU bs=24, your calculation is correct (~83.3k iterations).

No need to modify `train_iter`, and the code will automatically determine iterations based on the number of epochs.
